xdsl/dialects/csl.py

```diff
@@ -883,6 +908,437 @@ def verify_(self) -> None:
             raise VerifyException(f"{self.name} can only operate on mem1d_dsd type")
 
 
+FunctionSignatures = list[
+    tuple[AttrConstraint | Attribute | type[Attribute] | TypeVar, ...]
+]
+
+
+class BuiltinDsdOp(IRDLOperation, ABC):
+    ops = var_operand_def()
+
+    SIGNATURES: ClassVar[FunctionSignatures]
+
+    def verify_(self) -> None:
+        def typcheck(
+            op_typ: Attribute,
+            sig_typ: AttrConstraint | Attribute | type[Attribute] | TypeVar,
+        ) -> bool:
+            if isinstance(sig_typ, type):
+                return isinstance(op_typ, sig_typ)
+            else:
+                return op_typ == sig_typ
+
+        for sig in self.SIGNATURES:
+            if len(self.ops) == len(sig):
+                if all(typcheck(op.type, sig_t) for (op, sig_t) in zip(self.ops, sig)):
+                    return
+        raise VerifyException("Cannot find matching type signature")
```

Comment:
Can you add a unit test for this (just in python, not filecheck). I think this would increase confidence here

Reply:
Added a python test as suggested.